### ISSUE TITLE
FloatingPointError and UnsatError for some problems

### ISSUE THREAD

**Issue by dlshriver** [stanleybak/nnenum#3]
Hi Stanley,

I'm trying to verify a small network with nnenum, but I'm running into a couple errors that I was hoping you could help me with. I've included a tar archive to help reproduce the errors ([repro.tar.gz](https://github.com/stanleybak/nnenum/files/6556367/repro.tar.gz)). I am trying the check a simple property of a network that takes in 2 inputs, *a* and *b*, and predicts class 0 if *a* is greater than *b*, and class 1 if *a* is less than *b*. The property specifies that when *a* is in the interval [-2.6, 3.4] and *b* is in the interval [-4.1, 1.9], then output 0 is less than output 1. I specify a less than comparison by setting the bias in the output constraints to -5e-324. I would expect nnenum to report that this network and property are unsafe and return a counter-example, however it returns a FloatingPointError:

```
Traceback (most recent call last):
  File "/tmp/nnenum_debug/nnenum/src/nnenum/enumerate.py", line 563, in worker_func
    w.main_loop()
  File "/tmp/nnenum_debug/nnenum/src/nnenum/worker.py", line 46, in main_loop
    self.finished_star() # this sets self.priv.ss to None
  File "/tmp/nnenum_debug/nnenum/src/nnenum/worker.py", line 826, in finished_star
    violation_star = self.shared.spec.get_violation_star(ss.star, safe_spec_list=ss.safe_spec_list)
  File "/tmp/nnenum_debug/nnenum/src/nnenum/specification.py", line 232, in get_violation_star
    lpi.add_dense_row(row, rhs, normalize=normalize)
  File "/tmp/nnenum_debug/nnenum/src/nnenum/lpinstance.py", line 493, in add_dense_row
    rhs = rhs / norm
FloatingPointError: underflow encountered in double_scalars
```

This can be replicated by running the script `runner.py` in the included archive as follows (setting `NNENUM_ROOT` as appropriate):

```
$ OPENBLAS_NUM_THREADS=1 OMP_NUM_THREADS=1 NNENUM_ROOT=/path/to/nnenum ./runner_simplified.py network.onnx
```

I've also tried to verify the same property, but using a value of 0 as the bias on the output constraint, rather than -5e-324, which resulted in a different error message. I can't tell if they have the same root cause. This can be reproduced by commenting out line 42 in `runner.py` and uncommenting line 43, and then re-running the script as above. For me this produces the following output:

```
Doing quick overapprox with 2 rounds...
Overapprox Round 1/2 has 1 set(s)
Layer 1/1: LinearOnnxSubnetworkLayer... 0.0 sec
Overapprox Round 2/2 has 3 set(s)
Layer 1/1: LinearOnnxSubnetworkLayer... 0.001 sec
Running in parallel with 8 processes

Worker 0 found a confirmed counterexample (concrete)

Unsafe Branch: ++ (Branch Mode: 0)
Note: minimize failed with fail_on_unsat was true, trying to reset basis...
still unsat after reset basis, trying no-dir optimization
Error: No-dir result was also infeasible!
Lp has 2 columns (variables) and 2 rows (constraints)
       *i0    *i1
min 0.0000 0.0000
subject to:
       BS     NU
 1: BS -0.707 0.7071  <= -0.000
 2: NU 0.7071 -0.707  <= -0.000
(*) Bounded variables:
i0 in [-2.5572571754455566, 3.4427428245544434]
i1 in [-4.0600762367248535, 1.939923882484436]


Traceback (most recent call last):
  File "nnenum/src/nnenum/enumerate.py", line 563, in worker_func
    w.main_loop()
  File "nnenum/src/nnenum/worker.py", line 46, in main_loop
    self.finished_star() # this sets self.priv.ss to None
  File "nnenum/src/nnenum/worker.py", line 831, in finished_star
    concrete_io_tuple = self.find_concrete_io(violation_star, ss.branch_tuples)
  File "nnenum/src/nnenum/worker.py", line 756, in find_concrete_io
    res = star_copy.minimize_vec(None, return_io=True)
  File "nnenum/src/nnenum/lp_star.py", line 505, in minimize_vec
    lp_result = self.lpi.minimize(lp_vec)
  File "nnenum/src/nnenum/lpinstance.py", line 730, in minimize
    raise UnsatError("minimize returned UNSAT and fail_on_unsat was True")
nnenum.lpinstance.UnsatError: minimize returned UNSAT and fail_on_unsat was True

Worker 6 had exception


Exception occured during execution
error
```

Do you have any suggestions on how to fix these errors?

Thank you,
David

**Comment by stanleybak** [comment]
Thanks for the detailed report.

Numerical LP's guaranteed accuracy in GLPK is around `1e-6` or `1e-7`, so using `-5e-324` is probably not going to be much difference. The underflow is reflecting this, where the constraints are normalized before inserting into the LP, which requires divide the right hand side (`-5e-324`) by the norm of the left-hand side. I pushed a commit to the repository where I added explicit control for underflow behavior. You can use one of the following settings to prevent errors from stopping the computation:

```
Settings.UNDERFLOW_BEHAVIOR = 'warn'
Settings.UNDERFLOW_BEHAVIOR = 'ignore'
```

You also did find a bug in some untested code for generating concrete counter-examples where the LP becomes infeasible (this shouldn't be a failure). This is fixed in the newest commit as well. Let me know if this works for you or if you still see problems.

**Comment by dlshriver** [comment]
Thank you! That commit seems to have fixed those issues, but I'm running into a different error now in the same function as the FloatingPointError. I have create a small script to reproduce the error [here (repro2.tar.gz)](https://github.com/stanleybak/nnenum/files/6571849/repro2.tar.gz)
For this one, I am using a small network that always outputs a value of 0, and I specify that the output must be greater than or equal to 1. I would expect nnenum to report that this network is unsafe given this property but running:

```bash
$ OPENBLAS_NUM_THREADS=1 OMP_NUM_THREADS=1 NNENUM_ROOT=/path/to/nnenum ./runner2.py const_zero.onnx
```

I get the output:

```
Doing quick overapprox with 2 rounds...
Overapprox Round 1/2 has 1 set(s)
Overapprox Round 2/2 has 3 set(s)
Running in parallel with 8 processes


Traceback (most recent call last):
  File "/tmp/nnenum_debug/nnenum/src/nnenum/enumerate.py", line 563, in worker_func
    w.main_loop()
  File "/tmp/nnenum_debug/nnenum/src/nnenum/worker.py", line 46, in main_loop
    self.finished_star() # this sets self.priv.ss to None
  File "/tmp/nnenum_debug/nnenum/src/nnenum/worker.py", line 825, in finished_star
    violation_star = self.shared.spec.get_violation_star(ss.star, safe_spec_list=ss.safe_spec_list)
  File "/tmp/nnenum_debug/nnenum/src/nnenum/specification.py", line 232, in get_violation_star
    lpi.add_dense_row(row, rhs, normalize=normalize)
  File "/tmp/nnenum_debug/nnenum/src/nnenum/lpinstance.py", line 490, in add_dense_row
    assert norm > 0
AssertionError

Worker 0 had exception
Worker 0 quitting due to exception in some worker
Exception occured during execution
error
```

Is this related to the previous issue?

**Comment by stanleybak** [comment]
Here, the issue was again with the normalization of constraints, which I believe was done to help with LP stability. The actual constraint you're adding now has all zeros, so the norm is actually zero so we can't divide by that. I pushed another commit that checks for this case and skips normalization if so. I also added another option to the settings to disable this manually if it continues to cause problems:

```Settings.SKIP_CONSTRAINT_NORMALIZATION = True```

Let me know if you still have trouble.
